---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# MultiRNAflow

<!-- badges: start -->
<!-- badges: end -->

Our R package MultiRNAflow, built from the R package DESeq2 \cite{Love2014DESeq2}, provides an easy to use unified framework allowing to automatically make both unsupervised and supervised (DE) analysis for datasets with an arbitrary number of biological conditions and time points.
Specifically, our package realizes:

\begin{enumerate}
\item Exploratory (unsupervised) analyses of the data.
\item Statistical (supervised) analysis of transcriptional response (differentially expressed (DE) genes).
\item Analysis of temporal expression profiles of DE genes from different groups.
\item Functional analysis of temporal expression clusters.
\end{enumerate}

In particular, our code makes a deep downstream analysis of DE information, e.g. identifying temporal patterns across biological conditions and DE genes which are specific to a biological condition for each time.

Our package takes as input a transcriptional dataset with an experimental design that includes multiple groups of individuals and/or multiple times, t0 (reference time) and ti (1 $\leq$ i $\leq$ n). The dataset is a table of raw counts where lines correspond to genes and columns correspond to samples. Each sample show the raw counts of an individual sequencing, corresponding to a biological condition, an individual sampling in this biological condition and a time.

## Installation

Before installing the necessary packages, you must install (or update) the R version 4.2.1 "Funny-Looking Kid" (released on 2022/06/23) from \href{https://cran.r-project.org}{CRAN (Comprehensive R Archive Network)}\newline

Then, in order to use the MultiRNAflow package, the following R packages must be installed:

\begin{itemize}
\item From \href{https://cran.r-project.org}{CRAN}:
\href{https://cran.r-project.org/web/packages/scales/index.html}{scales} (>= 1.2.1),
\href{https://cran.r-project.org/web/packages/ggsci/index.html}{ggsci} (>= 2.9),
\href{https://cran.r-project.org/web/packages/RColorBrewer/index.html}{RColorBrewer} (>= 1.1.3),
\href{https://cran.r-project.org/web/packages/reshape2/index.html}{reshape2} (>= 1.4.4),
\href{https://cran.r-project.org/web/packages/plyr/index.html}{plyr} (>= 1.8.8),
\href{https://ggplot2.tidyverse.org}{ggplot2} (>= 3.4.0),
\href{https://cran.r-project.org/web/packages/ggalluvial/index.html}{ggalluvial} (>= 0.12.3),
\href{https://cran.r-project.org/web/packages/ggrepel/index.html}{ggrepel} (>= 0.9.2),
\href{https://cran.r-project.org/web/packages/FactoMineR/index.html}{FactoMineR} (>= 2.6),
\href{https://cran.r-project.org/web/packages/factoextra/index.html}{factoextra} (>= 1.0.7),
\href{https://cran.r-project.org/web/packages/plot3D/index.html}{plot3D} (>= 1.4),
\href{https://cran.r-project.org/web/packages/plot3Drgl/index.html}{plot3Drgl} (>= 1.0.3),
\href{https://cran.r-project.org/web/packages/UpSetR/index.html}{UpSetR} (>= 1.4.0),
\href{https://cran.r-project.org/web/packages/gprofiler2/index.html}{gprofiler2} (>= 0.2.1).
%
\item From \href{https://cran.r-project.org}{CRAN} and usually already included by default in R:
graphics (>= 4.2.2), grDevices (>= 4.2.2), grid (>= 4.2.2), methods (>= 4.2.2), stats (>= 4.2.2), utils (>= 4.2.2).
%
\item From \href{https://bioconductor.org}{Bioconductor}:
\href{https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html}{SummarizedExperiment} (>= 1.28.0),
\href{https://bioconductor.org/packages/release/bioc/html/DESeq2.html}{DESeq2} (>= 1.38.1),
\href{https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html}{ComplexHeatmap} (>= 2.14.0),
\href{https://www.bioconductor.org/packages/release/bioc/html/Mfuzz.html}{Mfuzz} (>= 2.58.0).
%
\end{itemize}


Before installing a package, for instance the package FactoMineR, the user must check if the package is already installed with the command \texttt{library(FactoMineR)}. If the package has not been previously installed, the user must use the command \texttt{install.packages("FactoMineR")} (for CRAN). For beginners in programming, we recommend to follow the steps below for importing CRAN and Bioconductor packages.

For the packages which must be download from CRAN,

```{r CranPck, eval=TRUE, echo=TRUE}
Cran.pck<-c("scales", "reshape2", "plyr",
            "ggplot2", "ggrepel", "ggalluvial",
            "FactoMineR", "factoextra",
            "plot3D", "plot3Drgl",
            "UpSetR", "gprofiler2")
```

the user can copy and paste the following lines of code for each package in order to download the missing packages.

```{rinstallCranPck, eval=FALSE, echo=TRUE}
Select.package.CRAN<-"FactoMineR"
#
if(!require(package=Select.package.CRAN,
            quietly=TRUE, character.only=TRUE, warn.conflicts=FALSE)){
  install.packages(pkgs=Select.package.CRAN, dependencies=TRUE)
}# if(!require(package=Cran.pck[i], quietly=TRUE, character.only=TRUE))
```

If the package is already installed (for instance here "FactoMineR"), the previous lines of code will return nothing.

For the packages which must be download from Bioconductor,

```{rBioconductorPck, eval=TRUE, echo=TRUE}
Bioconductor.pck<-c("Mfuzz", "SummarizedExperiment", "DESeq2", "ComplexHeatmap")
```

the user must first copy and paste the following lines of code in order to install "BiocManager"

```{r BiocManagerPck, eval=FALSE, echo=TRUE}
if(!require(package="BiocManager",
            quietly=TRUE, character.only=TRUE, warn.conflicts=FALSE)){
  install.packages("BiocManager")
}# if(!require(package="BiocManager", quietly=TRUE, character.only=TRUE))
```

then copy and paste the following lines of code in order to install the version 3.16 of bioconductor (it works with R version 4.2.0)

```{rBiocconductorVersion, eval=FALSE, echo=TRUE}
BiocManager::install(version = "3.16")
```

and then copy and paste the following lines of code for each package in order to download the missing packages.

```{r installBioconductorpck, eval=FALSE, echo=TRUE}
Select.package.CRAN<-"DESeq2"
if(!require(package=Select.package.CRAN,
            quietly=TRUE, character.only=TRUE, warn.conflicts=FALSE)){
  BiocManager::install(pkgs=Select.package.CRAN)
}# if(!require(package=Select.package.CRAN, quietly=TRUE, character.only=TRUE))
```

If the package is already installed (for instance here "DESeq2"), the previous lines of code will return nothing.

Once all packages have been installed, you may install and load our package.

```{r library, eval=TRUE, echo=TRUE, message=FALSE}
# BiocManager::install(pkgs="MultiRNAflow") # to install
library(MultiRNAflow)
```
