---
title: "Running `MultiRNAflow` on a RNA-seq raw counts with different time points and several biological conditions"
shorttitle: "MultiRNAflow"
author:
- name: "Rodolphe Loubaton"
  affiliation: "Université de Lorraine, CNRS, Inria, IECL, F-54000 Nancy, France"
- name: "Nicolas Champagnat"
  affiliation: "Université de Lorraine, CNRS, Inria, IECL, F-54000 Nancy, France"
- name: "Pierre Vallois"
  affiliation: "Université de Lorraine, CNRS, Inria, IECL, F-54000 Nancy, France"
- name: "Laurent Vallat"
  affiliation:
  - "Université de Strasbourg, INSERM, IRFAC UMR-S1113, Strasbourg, France"
  - "Department of molecular genetic of cancers, Molecular hematology unit, Strasbourg University Hospital, Strasbourg, France"
date: "`r format(Sys.Date(), '%d/%m/%Y')`"
output: 
  BiocStyle::html_document:
    toc: true
    number_sections: true
    toc-title: "INDICE"
bibliography: MultiRNAflowBiblio.bib
vignette: >
  %\VignetteIndexEntry{Running_analysis_with_MultiRNAflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE, results="hide"}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy = FALSE,
  cache = FALSE,
  #dev = "cairo_pdf",
  message = FALSE, error = FALSE, warning = FALSE
)
```


# Introduction

## Quick description of the document

This document is a quick workflow describing how to use our R package MultiRNAflow on one dataset (see [Dataset used as example]).\newline
For a more complete description of our package and complete outputs with graphs, the user must read our pdf file entitled 'MultiRNAflow_vignette.pdf".

## Dataset used as example

The **Mouse dataset 2** [@Weger2021TemporalMouseLiver] is accessible on the Gene Expression Omnibus (GEO) database with the accession number GSE135898.

This dataset contains the temporal transcription profile of 16 mice with Bmal1 and Cry1/2 knocked-down under an ad libitum (AL) or night restricted feeding (RF) regimen. Data were collected at 0, 4h, 8h, 16, 20h and 24h. Therefore, there are six time points and eight biological conditions. As there are only two mice per biological condition, we decided not to take into account the effect of the regimen. The dataset contains temporal expression data of 40327 genes. \newline

To illustrate the use of our package, we take 500 genes, over the global 40327 genes in the original dataset.
This sub dataset is saved in the file **RawCounts_Weger2021_MOUSEsub500**.


# Quick workflow

## Load package and example dataset

1. Load MultiRNAflow

```{r library, warning=FALSE, message=FALSE}
library(MultiRNAflow)
```

2. Load **Mouse dataset 2**

```{r DataWeger2021}
data("RawCounts_Weger2021_MOUSEsub500")
```

## Exploratory data analysis

3. Normalization with DATAnormalization()

```{r NormalizationMouse2, eval=TRUE, results='hide', fig.show='hide'}
res.norm<-DATAnormalization(RawCounts=RawCounts_Weger2021_MOUSEsub500,
                            Column.gene=1,
                            Group.position=1,
                            Time.position=2,
                            Individual.position=3,
                            Normalization="vst",
                            Blind.rlog.vst=FALSE,
                            Plot.Boxplot=TRUE,
                            Colored.By.Factors=TRUE,
                            Color.Group=NULL,
                            path.result=NULL)
```

Write `?DATAnormalization` in your console for more information about the function.

4. Principal component analysis (PCA) with PCAanalysis()

```{r PCAMus2, eval=TRUE, results='hide', fig.show='hide'}
res.pca<-PCAanalysis(ExprData=res.norm$NormalizedData,
                     Column.gene=1,
                     Group.position=1,
                     Time.position=2,
                     Individual.position=3,
                     Mean.Accross.Time=FALSE,
                     gene.deletion=NULL,
                     sample.deletion=NULL,
                     Supp.del.sample=FALSE,
                     Color.Group=NULL,
                     Cex.label=0.6,
                     Cex.point=0.7, epsilon=0.2,
                     Phi=25,Theta=140,
                     D3.mouvement=FALSE,
                     path.result=NULL,
                     Name.folder.pca=NULL)
```

Write `?PCAanalysis` in your console for more information about the function.


5. Hierarchical Clustering on Principle Components (HCPC) with HCPCanalysis()

```{r HCPCmus2500, warning=FALSE, message=FALSE, eval=FALSE}
res.hcpc<-HCPCanalysis(ExprData=res.norm$NormalizedData,
                       Column.gene=1,
                       Group.position=1,
                       Time.position=2,
                       Individual.position=3,
                       gene.deletion=NULL,
                       sample.deletion=NULL,
                       Supp.del.sample=FALSE,
                       Phi=25,Theta=140,
                       Cex.point=0.6,
                       epsilon=0.2,
                       Cex.label=0.6,
                       D3.mouvement=FALSE,
                       path.result=NULL,#FolderResultsLeuk,
                       Name.folder.hcpc=NULL)
```

Write `?HCPCanalysis` in your console for more information about the function.

6. Temporal clustering analysis with MFUZZanalysis().

```{r MfuzzMus2, eval=TRUE, results='hide', fig.show='hide'}
res.mfuzz<-MFUZZanalysis(ExprData=res.norm$NormalizedData,
                         Column.gene=1,
                         Group.position=1,
                         Time.position=2,
                         Individual.position=3,
                         DataNumberCluster=NULL,
                         Method="hcpc",
                         Membership=0.5,
                         Min.std=0.1,
                         path.result=NULL,
                         Name.folder.mfuzz=NULL)
```

Write `?MFUZZanalysis` in your console for more information about the function.

7. Plot temporal expression with with DATAplotExpressionGenes()

```{r DATAplotExpressionGenesMusBmCrKoWt, eval=TRUE, results='hide', fig.show='hide'}
res.expr<-DATAplotExpressionGenes(ExprData=res.norm$NormalizedData,
                                  Vector.row.gene=c(17),
                                  Column.gene=1,
                                  Group.position=1,
                                  Time.position=2,
                                  Individual.position=3,
                                  Color.Group=NULL,
                                  path.result=NULL)
```

Write `?DATAplotExpressionGenes` in your console for more information about the function.

## Supervised statistical analysis

8. Differential Expresion (DE) analysis with DEanalysisGlobal()

```{r DEMusBmCrKoWt, eval=FALSE, echo=TRUE}
res.de<-DEanalysisGlobal(RawCounts=RawCounts_Weger2021_MOUSEsub500,
                         Column.gene=1,
                         Group.position=1,
                         Time.position=2,
                         Individual.position=3,
                         pval.min=0.05,
                         pval.vect.t=c(0.05,0.05,0.05,0.05,0.05),
                         log.FC.min=1,
                         LRT.supp.info=FALSE,
                         path.result=NULL,
                         Name.folder.DE=NULL)
```

Write `?DEanalysisGlobal` in your console for more information about the function.

9. Volcano and ratio intensity (MA) plots with DEplotVolcanoMA()

```{r VolcanoMA_Mouse2, eval=FALSE, echo=TRUE}
res.VolcanoMA<-DEplotVolcanoMA(Res.DE.analysis=res.de,
                               NbGene.plotted=2,
                               SizeLabel=3,
                               Save.plots=TRUE)
```

Write `?DEplotVolcanoMA` in your console for more information about the function.

10. Heatmaps with DEplotHeatmaps()

```{r Heatmaps_Mouse2, eval=FALSE, echo=TRUE}
res.heatmaps<-DEplotHeatmaps(Res.DE.analysis=res.de,
                             ColumnsCriteria=20,
                             Set.Operation="union",
                             NbGene.analysis=20,
                             SizeLabelRows=5,
                             SizeLabelCols=5,
                             Save.plots=TRUE)
```

Write `?DEplotHeatmaps` in your console for more information about the function.

11. GO enrichment analysis with GSEAQuickAnalysis() and GSEApreprocessing()

```{r GSEAquickAnalysis_Mouse2, eval=FALSE}
res.gprofiler2<-GSEAQuickAnalysis(Res.DE.analysis=res.de,
                                  ColumnsCriteria=2,
                                  ColumnsLog2ordered=NULL,
                                  Set.Operation="union",
                                  Organism="mmusculus",
                                  MaxNumberGO=20,
                                  Background=FALSE,
                                  Save.plots=TRUE)
```

Write `?GSEAQuickAnalysis` in your console for more information about the function.

```{r GSEAprepro_Mouse2, eval=FALSE}
res.preprocessing<-GSEApreprocessing(Res.DE.analysis=res.de,
                                     ColumnsCriteria=2,
                                     Set.Operation="union",
                                     Save.files = TRUE)
```

Write `?GSEApreprocessing` in your console for more information about the function.

# SessionInfo

Bioconductor recommends and checks (see [here](https://bioconductor.org/packages/devel/bioc/vignettes/BiocCheck/inst/doc/BiocCheck.html)) that vignette code contains `sessionInfo()` for reproducibility.

```{r SessionInfo}
sessionInfo()
```


